File: Count Good Nodes in Binary Tree - Leetcode 1448/Count Good Nodes in Binary Tree - Leetcode 1448.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = val;
 *     this.left = left;
 *     this.right = right;
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number}
 */
var goodNodes = function(root) {
    if (!root) return 0;

    let goodNodes = 0;
    const stk = [[root, -Infinity]];

    while (stk.length > 0) {
        const [node, largest] = stk.pop();

        if (largest <= node.val) {
            goodNodes++;
        }

        const newLargest = Math.max(largest, node.val);
        if (node.right) stk.push([node.right, newLargest]);
        if (node.left) stk.push([node.left, newLargest]);
    }

    return goodNodes;
};

File: Count Good Nodes in Binary Tree - Leetcode 1448/Count Good Nodes in Binary Tree - Leetcode 1448.py
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        good_nodes = 0
        stk = [(root, float("-inf"))]

        while stk:
            node, largest = stk.pop()
            
            if largest <= node.val:
                good_nodes += 1

            largest = max(largest, node.val)
            if node.right: stk.append((node.right, largest))
            if node.left: stk.append((node.left, largest))

        return good_nodes

# Time Complexity: O(n)
# Space Complexity: O(n)
